### PR TITLE
Add Appveyor Windows testing

### LINKED ISSUE #80: Get appveyor CI for windows working
@rok-cesnovar I added an `.appveyor.yml` config file so we can try continuous integration on windows via appveyor (in addition to linux on travis). It's working (running on appveyor) but errorring when running some of the unit tests (https://ci.appveyor.com/project/jgabry/cmdstanr/history). 

I also started a branch `appveyor-ci-windows` for if we need to play around with changing things in the package (other than just the `.appveyor.yml` file) to get it to work. 

Most of the `.appveyor.yml` file is set to the defaults recommended in https://github.com/krlmlr/r-appveyor#readme, except I added to `build_script:` the line 

```r
  - cmd: bash inst/make_cmdstan.sh -j2 -w
```

which seems to at least run but may not be successful. 

One issue seems to be that the default paths used on appveyor may be different than what we're expecting in various functions (e.g., `make_cmdstan.sh`, `set_cmdstan_path()`). But I'm having a bit of trouble figuring out what's going on since Windows confuses the hell out of me!

The nice thing about appveyor though is that you can download the failure logs by going to the "Artifacts" tab when on the page for a particular build.

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Fixes #80 by adding support for testing on appveyor.
For now this builds the package and runs:

R CMD check with the flags: "--no-manual --as-cran --install-args=--build"

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, University of Ljubljana

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: .appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+skip_branch_with_pr: true
+version: '{build}'
+image: Visual Studio 2017
+
+platform:
+  - x64
+
+# cache:
```
Comment (rok-cesnovar) [inline .appveyor.yml]: I am not good with appveyor caching so not caching for now. I am not sure how to not install rtools and R when cached versions are found. But I guess its not that bad, the bottleneck is compiling cmdstan anyways.

File: .appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+environment:
+  NOT_CRAN: true
+  _R_CHECK_FORCE_SUGGESTS_: false
+  USE_RTOOLS: true
+  HOME: C:\Users\appveyor\
+  _R_VERSION: "3.6.1"
+  _RTOOLS_VERSION: "35"
+  _CMDSTAN_INSTALL_FLAGS: "-j2 -w"
+  _R_BUILD_FLAGS: "--no-manual"
+  _R_CHECK_FLAGS: "--no-manual --as-cran --install-args=--build"
```
Comment (rok-cesnovar) [inline .appveyor.yml]: Do we need other flags?

File: R/run.R [hunk 3]
```diff
@@ -423,9 +423,9 @@ CmdStanProcs <- R6::R6Class(
     },
     is_error_message = function(line) {
       startsWith(line, "Exception:") ||
-      (regexpr(line, "either mistyped or misplaced.") > 0) ||
-      (regexpr(line, "A method must be specified!") > 0) ||
-      (regexpr(line, "is not a valid value for") > 0)
+      (regexpr("either mistyped or misplaced.", line, perl = TRUE) > 0) ||
+      (regexpr("A method must be specified!", line, perl = TRUE) > 0) ||
+      (regexpr("is not a valid value for", line, perl = TRUE) > 0)
```
Comment (rok-cesnovar) [inline R/run.R]: This is a bug in the use of regexpr, the pattern and text were switched. I think this should now fix #69 but have to double check.

File: R/run.R [hunk 4]
```diff
@@ -438,22 +438,22 @@ CmdStanProcs <- R6::R6Class(
           last_section_start_time <- private$chain_info_[id,"last_section_start_time"]
           state <- private$chain_info_[id,"state"]
           next_state <- state
-          if (state == 1 && regexpr("Iteration:", line) > 0) {
+          if (state == 1 && regexpr("Iteration:", line, perl = TRUE) > 0) {
```
Comment (rok-cesnovar) [inline R/run.R]: Added perl = TRUE for handling the "\U" and "\a" characters in stdout.

File: tests/testthat/helper-envvars-and-paths.R [hunk 5]
```diff
@@ -22,14 +22,6 @@ not_on_cran <- function() {
...
 }
 
-set_cmdstan_path_for_tests <- function() {
-  if (on_appveyor()) {
-    set_cmdstan_path("C:/MinGW/msys/1.0/.cmdstanr/cmdstan")
-  } else {
-    set_cmdstan_path()
-  }
-}
-
```
Comment (rok-cesnovar) [inline tests/testthat/helper-envvars-and-paths.R]: This isnt needed anymore.

File: tests/testthat/test-model-compile.R [hunk 6]
```diff
@@ -37,13 +37,13 @@ test_that("compile() method forces recompilation if changes in flags", {
   skip_on_cran()
   skip_on_ci() # FIXME: this test seems to pass locally but fail on travis
   expect_message(
-    mod$compile(opencl=TRUE),
+    mod$compile(threads=TRUE),
```
Comment (rok-cesnovar) [inline tests/testthat/test-model-compile.R]: opencl isnt present on Windows by default. TBB is so I think testing this behavior with threads is better. I also think that is why Travis was failing, so `skip_on_ci()` might not be needed here.

File: tests/testthat/test-path.R [hunk 7]
```diff
@@ -53,7 +53,7 @@ test_that("cmdstanr_initialize() also looks for default path", {
 
   unset_cmdstan_path()
   cmdstanr_initialize()
-  expect_equal(cmdstan_path(), PATH)
+  expect_equal(tolower(cmdstan_path()), tolower(PATH))
```
Comment (rok-cesnovar) [inline tests/testthat/test-path.R]: Windows had issues with this test.

File: .appveyor.yml (hunk 1 above)
Comment (jgabry) [reply .appveyor.yml]: That's fine. We can maybe add caching later but not a big deal either way

File: .appveyor.yml (hunk 2 above)
Comment (jgabry) [reply .appveyor.yml]: I don't think so

File: tests/testthat/helper-envvars-and-paths.R (hunk 5 above)
Comment (jgabry) [reply tests/testthat/helper-envvars-and-paths.R]: Great!

File: tests/testthat/test-model-compile.R (hunk 6 above)
Comment (jgabry) [reply tests/testthat/test-model-compile.R]: Ok cool. I just pushed a commit removing `skip_on_ci()` so we'll see if travis can handle it!

**Comment by jgabry** [comment]
> Just adding a few comments to help reviewing.

Thanks those comments were super helpful!

I'll merge this after ci runs again (i just removed skip_on_ci() from that compilation test)

**Comment by rok-cesnovar** [comment]
Yeah, I never worked with appveyor before so it was a challenge for me :)

Now on to finishing the read_sample_csv. Should be ready tomorrow.

**Comment by jgabry** [comment]
> Now on to finishing the read_sample_csv. Should be ready tomorrow.

Awesome!

**Comment by rok-cesnovar** [comment]
There is something else with Travis I guess.

**Comment by jgabry** [comment]
Hmmm. travis didn't like the test for change in compiler flags, but the error is "cannot open the connection". So it doesn't seem to be coming from recompilation but rather from trying to write to `make_local_path`: 

```r
 ── 1. Error: compile() method forces recompilation if changes in flags (@test-mo
  cannot open the connection
  Backtrace:
   1. testthat::expect_message(mod$compile(threads = TRUE), "change in the compiler flags was found")
   6. mod$compile(threads = TRUE)
   7. cmdstanr:::set_make_local(...)
   8. base::writeLines(make_local_content, make_local_path)
   9. base::file(con, "w")
```

**Comment by jgabry** [comment]
Anyway, that's not a problem with your PR so I'll turn off this test on travis for now and merge this and then open a separate issue for figuring it out on travis

**Pull request merged**